src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        if (index == 0) {
+            T removedElement = (T) head.value;
+            head = head.next;
+            size--;
+            return removedElement;
+        } else if (index == size - 1) {
+            T removedElement = (T) tail.value;
+            tail = tail.prev;
+            size--;
+            return removedElement;
+        }
+        T removedElement;
+        Node<T> current = getNodeByIndex(index);
+        removedElement = (T) current.value;
+        unlink(current);
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value != null && current.value.equals(object) || current.value == object) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void outOfBoundsCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Wrong index: " + index);
+        }
+    }
+
+    private void unlink(Node current) {
+        if (current == head) {
+            head = head.next;
+        } else if (current == tail) {
+            tail = tail.prev;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current;
+        if (index > size / 2) {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        } else {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        }
+        return current;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        private Node(T value) {
```

Comment:
i think we should initialize all the fields in the constructor, that way the whole code will be a bit more readable

Reply:
agree